exist-core/src/main/java/org/exist/xquery/value/FunctionParameterFunctionSequenceType.java

```diff
@@ -0,0 +1,171 @@
+        super(attributeName, primaryType, cardinality, description);
+        this.parameters = parameterTypes;
+        this.arity = parameterTypes.length;
+    }
+
+    /**
+     * FIXME: is this constructor needed?
+     * @param attributeName
+     */
+    public FunctionParameterFunctionSequenceType(final String attributeName) {
+        super(attributeName);
+    }
+
+    @Override
+    public boolean checkType(Sequence seq) throws XPathException {
+        // all functions?
+        if (!Type.subTypeOf(seq.getItemType(), getPrimaryType())) {
+            throw new XPathException(ErrorCodes.XPTY0004,
+                    "Type error: expected type: "
+                            + Type.getTypeName(getPrimaryType())
+                            + "; got: "
+                            + Type.getTypeName(seq.getItemType()));
+        }
+
+        // check each ref
+        FunctionReference next;
+        for (final SequenceIterator i = seq.iterate(); i.hasNext(); ) {
+            next = (FunctionReference) i.nextItem();
+            if (!checkType(next)) {
+                throw new XPathException(ErrorCodes.XPTY0004,
+                        "Type error: expected type: "
+                                + Type.getTypeName(getPrimaryType())
+                                + "; got: "
+                                + Type.getTypeName(seq.getItemType()));
+            }
+        }
+        return true;
+    }
+
+    public boolean checkType(final FunctionReference ref) throws XPathException {
+        final FunctionSignature sig = ref.getSignature();
+        // check arity of referenced function call
+        arityMatches(sig);
+
+        // check return type
+        returnTypeMatches(sig);
+
+        if (arityUnspecified()) { return true; }
+
+        // check argumentTypes
+        final SequenceType[] arguments = sig.getArgumentTypes();
+        for (int i = 0; i < arguments.length; i++) {
+            final int argumentType = arguments[i].getPrimaryType();
+            final int parameterType = parameters[i].getPrimaryType();
+            // ITEM is likely unspecified return type - catch error later
+            if (argumentType != Type.ITEM && !Type.subTypeOf(argumentType, parameterType)) {
+                // throw
+                throw new XPathException(ErrorCodes.XPTY0004,
+                        "Type error: expected type: " + Type.getTypeName(parameterType)
+                                + "; got: " + Type.getTypeName(argumentType));
+            }
+        }
+        return true;
+    }
+
+    private void arityMatches (final FunctionSignature sig) throws XPathException {
+        if (arityUnspecified()) { return; }
+        final int otherArity;
+
+        if (sig.isOverloaded()) {
+            // concat#3 will not return the correct argument count but the number of
```

Comment:
I am not sure why this would be dealing with `concat#3`?
`fn:concat` is a *var-args* function; the only on in XQuery actually.